Added a new example for doing preferential batch bayesian optimization

esiivola opened the pull request:
*Issue #, if available:*

*Description of changes:*
Added an example for performing the experiments in https://arxiv.org/abs/2003.11435

See /emukit/examples/preferential_batch_bayesian_optimization/README.md for details.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

apaleyes commented on emukit/examples/preferential_batch_bayesian_optimization/pbbo/inferences/sexpgp_comparison.stan:
```diff
@@ -0,0 +1,56 @@
+functions {
```
Can you briefly explain why this bit has to be implemented in Stan? not saying this is a no-go, just trying to understand why we suddenly need it in otherwise Python-only codebase

esiivola replied:
It is a baseline method we compare our method against. Implementing the method on Stan is fast and the code is much more readable than making an implementation from scratch.

apaleyes replied:
ok, fair enough. in that case can you add a comment at the top of this file with the same explanation?

esiivola commented:
Hi! Thanks for the fast reply! Yes, I can make a dependency installation instead, no problem!

esiivola commented:
I

> Ok, so we have some code in here which is under MIT license, and some under BSD. My understanding is that if you copy such code, you need to include corresponding licenses, and I doubt we can do that with emukit.
> 
> But it is ok to use software under these licenses. Can the implementation be done with dependency installation instead?

This is now done!

esiivola commented:
> Ok, so we have some code in here which is under MIT license, and some under BSD. My understanding is that if you copy such code, you need to include corresponding licenses, and I doubt we can do that with emukit.
> 
> But it is ok to use software under these licenses. Can the implementation be done with dependency installation instead?

Hi! This is done now. I moved the black box functions to a separate git repo and also added a dependency for stan_utility.

apaleyes commented on emukit/examples/preferential_batch_bayesian_optimization/pbbo/gp_models.py:
```diff
@@ -0,0 +1,468 @@
+import numpy as np
+
+import matplotlib
+matplotlib.use('Agg')
```
Is it fixing the same thing that installation is supposed to fix on OSX? If so, let's just leave it there, so that code itself doesn't force anything. If people found an alternative way to fix the issue, or didn't have the issue in the first place, hardcoding the backend may cause more harm than good.

esiivola replied:
We actually don't need matplotlib at all so we can also remove this. Thanks for noticing!

apaleyes commented on emukit/examples/preferential_batch_bayesian_optimization/pbbo/inferences/mcmc_batch_comparison.py:
```diff
@@ -0,0 +1,132 @@
+import sys
+
+from typing import Tuple, List, Callable, Dict
+# set environmental variable STAN_NUM_THREADS
+# Use 4 cores per chain
+os.environ['STAN_NUM_THREADS'] = "1" # To make our lives easier on clusters
```
this should probably be moved to installation/running instructions

esiivola replied:
Yes, good idea!

apaleyes commented on emukit/examples/preferential_batch_bayesian_optimization/pbbo/util.py:
```diff
@@ -0,0 +1,291 @@
+from typing import Callable, List, Tuple, Dict
+
+'''
+nearestPD and isPD are copied from https://github.com/alan-turing-institute/bocpdms with the following licence:
+
+
+MIT License
```
This file should go as well, we cannot add MIT license to emukit at this time

esiivola replied:
Definitely! My bad for not noticing this! Sorry!

esiivola commented on emukit/examples/preferential_batch_bayesian_optimization/installation.txt:
```diff
@@ -0,0 +1,12 @@
+
+#On macs matplotlib doesn't necessarily work so running the following lines should fix it
+mkdir -p ~/.matplotlib
+echo "backend: TkAgg" > ~/.matplotlib/matplotlibrc
```
We actually don't need matplotlib so i'll remove this

esiivola commented on emukit/examples/preferential_batch_bayesian_optimization/installation.txt:
```diff
@@ -0,0 +1,12 @@
+
+# First install the needed basic libraries
+conda install pip numpy scipy=1.1.0 cython matplotlib
+pip install GPy numdifftools pystan quadpy GPyOpt gitpython pandas emukit pyDOE stan_utility git+https://github.com/esiivola/evalset@master
```
Good spot! You are right!

apaleyes commented on emukit/examples/preferential_batch_bayesian_optimization/README.md:
```diff
@@ -0,0 +1,30 @@
+
+- "acquisitions" folder contain the implementations for the acquisition functions.
+
+- sigopt_functions.py contains all the functions used as examples in the paper.
+  The functions are mostly copied from
+  https://github.com/sigopt/evalset/blob/master/evalset/test_funcs.py
```
Looks like this file is gone now

esiivola replied:
Right, I'll fix this as well!
